Update core.py - 'params' & mp for custom strategies

DrPaprikaa opened the pull request:
Hi @twopirllc, 

Here are the changes I propose to allow for the use of **mp** in CustomStrategies, as well as specifying the parameters in `ta` with `'params'`, in the form :

```python
CustomStrategy = ta.Strategy(
    name="MyStrat",
    ta=[
        {'kind':'sma', 'params':(30)}, 
        {'kind':'bbands', 'params':(20,2)},
        {'kind':'macd', 'params':(8,21)}
    ]
)
```
I have accounted for the case where some indicators are specified with `'params'` and the others with their explicit `fast`, `slow`..., and for the case of chained indicators, for example :

```python
CustomStrategy = ta.Strategy(
    name="MyStrat",
    ta=[
        {"kind":"ema",  "length": 5}, 
        {'kind':'bbands', 'params':(20,2)},
        {"kind":"linreg", "close": "EMA_5", "length": 8, "prefix": "EMA_5"}
        {"kind":"macd", "fast":9, "slow":19, "signal":7},
    ]
)
``` 

What I did :


1. deleted the `finalize()` method
2. re-writted **all** the indicators calls without `finalize()`. Deleted `open`, `high`, `low`, `close` & `volume` if present in the definition. The use of `close = self._get_column(kwargs.pop('close', 'close'))` allows for chained indicators via `pop()`, with a default on the actual `close` column of the df (respectively `open`, `high`...)
3. moved `mp_worker()` inside the `AnalysisIndicators` class and re-writed it consequently
4. writted a mp ordered map pool for CustomStrategies and skrinked the code
5. writted the same code without mp, to allow for comparison.

If I'm not mistaken these changes don't break any compatibility on the user's end.

Take your time to review the changes, you'll maybe see some details that I didn't !

Thanks,
DrPaprikaa

twopirllc commented:
Hey @DrPaprikaa,

I will play around with it soon and test it on the _development_ branch..

Thanks for taking the time to contribute to these issues and improving **Pandas TA**!

Thanks,
KJ

DrPaprikaa commented:
No problem, happy to contribute to this awesome project !

DrPaprikaa

twopirllc commented:
@DrPaprikaa,

The ```strategy``` upgrade works well. No issues so far with it. 😎 

However was the ```finalize``` decorator interfering with ```strategy```?

Because it's removal has made it unable to ```append``` or add ```prefix/suffix``` to the the ```df``` DataFrame when doing:
```python
ao = df.ta.ao(append=True, prefix="pre", suffix="suff")
```

It is a simple fix to append within a method to get it working like it was originally before writing ```finalize```. Such as:
```python
    def ao(self, fast=None, slow=None, offset=None, **kwargs):
        high  = self._get_column(kwargs.pop("high", "high"))
        low   = self._get_column(kwargs.pop("low", "low"))

        result = ao(high=high, low=low, fast=fast, slow=slow, offset=offset, **kwargs)
        self._add_prefix_suffix(result=result, **kwargs)
        self._append(result=result, **kwargs)
        return result
```

Will re implementing ```finalize``` break ```strategy```? It is an easy thing to check and try for me. Just curious if ```finalize``` affected ```strategy``` in your development?

Thanks,
KJ

DrPaprikaa commented:
Hi @twopirllc,

Sure let me try it and do some tests. I also spotted typos here and there, so I'll use the occasion to correct them.

DrPaprikaa

twopirllc commented:
@DrPaprikaa,

Sounds good. Also doing some refactoring on my end.

Thanks,
KJ

twopirllc commented:
@DrPaprikaa,

So most of it is ready to go! Did some refactoring, core documenting, and now _all_ the tests are passing. Just left is some documentation and notebook updating with an example or two.

DrPaprikaa commented:
@twopirllc,

I didn't find a way to correctly use `finalize()`, so I added back `add_prefix()` and `append()` to the indicators calls, so it now behaves as intended.

In the case where an indicator returns a `df`, I added an option, via `kwargs`, to append only one desired column to the data. It goes like this :
```
CustomStrategy = ta.Strategy(
    name="MyStrat",
    ta=[
        {'kind':'bbands', 'params':(20,), 'col_number':2},
    ]
)
```
or alternatively : `ao = df.ta.bbands(append=True, col_number=2)`

which can be usefull if the user needs only the upper bband for example.

Btw I have no idea how to use Github Desktop yet, so please discard commits 2 & 4 ! My 3rd commit is the one to go with, but it does not include the code refractoring that you did (where can I see what you wrote btw ?)

DrPaprikaa

DrPaprikaa commented:
> Might want to try and figure out how to update your branch with the _development_ branch make the `col_number` work from that point. 

If you're happy with the `col_number`  solution I suggested, then updating the code is as simple as copy-pasting the line `result = result.iloc[:,int(kwargs['col_number'])] if 'col_number' in kwargs and kwargs['col_number'] is not None else result` in the corresponding indicators calls.

Don't hesitate if I can be of any help. Looking forward to seeing the result !

In the future should I make pull requests on the master or dev branch ?
DrPaprikaa

twopirllc commented:
@DrPaprikaa,

If it may be a useful feature, then adding ```col_number``` but will not know unless it is implemented. Can you only target one column or more than one?

> Don't hesitate if I can be of any help. Looking forward to seeing the result !

From what I have seen so far, you are **welcome** to help contribute and knock out any issues on the issue board.

> In the future should I make pull requests on the master or dev branch ?

I push new stuff onto the _development_  branch and critical **bugs** onto _master_. But yeah, try to make use of the _dev_ branch.

KJ

DrPaprikaa commented:
@twopirllc, 
> Adding `col_number` may be a useful feature, but we will not know unless it is implemented.

I've already implemented it in commit 3, it works well.

> Can you only target one column or more than one?

 Only one in the current form. Updating to multiple columns would be done by changing the line:

`result = result.iloc[:,int(kwargs['col_number'])] if 'col_number' ...`

to something like:

`result = result.iloc[:,[int(n) for n in kwargs['col_number']]] if 'col_number' ...`

and by specifying `'col_numbers'` as a tuple. I haven't tested it though.

DrPaprikaa

twopirllc commented:
@DrPaprikaa,

Last night and since the removal of ```finalize```, I created a private method ```_post_process()``` to handle ```_append``` and ```_add_prefix_suffix``` before returning the ```result```.

Sample indicator definition:
```python
    def bias(self, length=None, mamode=None, offset=None, **kwargs):
        close  = self._get_column(kwargs.pop("close", "close"))
        result = bias(close=close, length=length, mamode=mamode, offset=offset, **kwargs)
        return self._post_process(result, **kwargs)
```

As for ```col_number``` lets do a seperate pull request from _development_ so everything is insync again and close this one out.

>     Adding col_number may be a useful feature, but we will not know unless it is implemented.
> I've already implemented it in commit 3, it works well.
>     Can you only target one column or more than one?
> Only one in the current form. Updating to multiple columns would be done by changing the line:
> result = result.iloc[:,int(kwargs['col_number'])] if 'col_number' ...
> to something like:
> result = result.iloc[:,[int(n) for n in kwargs['col_number']]] if 'col_number' ...
> and by specifying 'col_numbers' as a tuple. I haven't tested it though.

KJ

DrPaprikaa commented:
Sure I'm closing this pull request, moving on to `'col_number'` :)
